### PR TITLE
MQTT Implementation and HiveMQ connection

### LINKED ISSUE #43: Transform subscription information into MQTT message
DoD:
- [ ] OnOff subscription information gets transformed into JSON
- [ ] Resulting JSON is sent over MQTT

### PR DESCRIPTION
**Pull request by Ti-ger**
Closes #43 
# Pull Request Checklist
## Author

- [x] Built code and tested locally
- [x] Removed any TODOs, etc.
- [x] Metioned the issue this PR closes
- [x] Good description of what the PR does written

## Reviewer

- [ ] Built and tested code locally
- [ ] Fully understood the PR
- [ ] All open conversations resolved

### CODE REVIEW

**Comment by Ti-ger** [comment]
Adds MQTT Support by using the ESP-IDF MQTT Client. 
Also implements the CLI and some logic to delay commands when the Connection HiveMQ is disrupted.
TODOs:
- Check if we really want to use this tcp client and this event-loop. 
My guess is: this is exactly what Matter uses under the hood, but we should verify this.

- Create a shared HiveMQ Account
- Store credentials in a way not everybody can see them...

**Comment by vringar** [comment]
Just curious, why is this not implemented via another git submodule?
This seems to be a copy of a whole repository including Travis and Gitlab CI config?

**Comment by Ti-ger** [comment]
> 

You are right, this is a leftover. I thought we may have to replace all the socket calls with "Matter Socket" calls and so on. 
Luckily this was not necessary. I will remove all the files, thanks for the hint!

**Comment by bowski23** [comment]
Don't worry about the force-pushed changes. This was done to get rid of the secrets stored in past commits.

File: src/main/CMakeLists.txt [hunk 1]
```diff
@@ -9,6 +9,7 @@ idf_component_register(PRIV_INCLUDE_DIRS
                       "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/zzz_generated/light-switch-app"
                       "${CMAKE_CURRENT_LIST_DIR}/include"
                       "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/examples/platform/esp32"
+                      "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/zzz_generated/app-common/app-common/zap-generated/ids"
```
Comment (vringar) [inline src/main/CMakeLists.txt:12]: Did I introduce this artifact or is this deliberate?

File: src/main/Kconfig.projbuild [hunk 2]
```diff
@@ -73,3 +73,11 @@ menu "Demo"
...
 
 endmenu
+menu "MQTT Configuration"
+    config MQTT_CLIENT_URI
+       string "MQTT URI"
+    config MQTT_CLIENT_USERNAME
+       string "MQTT Username"
+    config MQTT_CLIENT_PASSWORD
+	string "MQTT Password"
+endmenu
```
Comment (vringar) [inline src/main/Kconfig.projbuild:83]: Should we mention this in the readme?

File: src/main/MQTTManager.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
+#include "MQTTManager.h"
```
Comment (vringar) [inline src/main/MQTTManager.cpp:1]: License header missing

File: src/main/MQTTManager.cpp [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    default:
+        ChipLogError(NotSpecified, "MQTTCommandWorkerFunction - Invalid Task");
+        break;
+    }
+}
+
+//
+// Note: this function is for testing purposes only publishing part of the active partition
+//       (to be checked against the original binary)
+//
```
Comment (vringar) [inline src/main/MQTTManager.cpp:95]: I don't understand this comment

File: src/main/include/MQTTManager.h [hunk 5]
```diff
@@ -0,0 +1,31 @@
+#pragma once
```
Comment (vringar) [inline src/main/include/MQTTManager.h:1]: License header missing

File: src/main/include/Subscription.h [hunk 6]
```diff
@@ -5,16 +5,21 @@
 #pragma once
```
Comment (vringar) [inline src/main/include/Subscription.h:5]: License header missing

File: src/main/include/Subscription.h [hunk 7]
```diff
@@ -39,20 +46,80 @@ class Subscription : public InteractionModelReports, public chip::app::ReadClien
...
+        ESP_LOGI("Subscription", "Fabric: %llu", mfabricId);
 
         if (data == nullptr)
         {
             ChipLogError(chipTool, "Response Failure: No Data");
             mError = CHIP_ERROR_INTERNAL;
             return;
         }
+
+        // TODO: What we want to use:
```
Comment (vringar) [inline src/main/include/Subscription.h:70]: Is this todo for this PR or for followup?

**Review by vringar — CHANGES_REQUESTED** [review]
This is mostly a first pass. I need to look through this more thoroughly.

File: src/main/MQTTManager.cpp (hunk 4 above)
Comment (Ti-ger) [reply src/main/MQTTManager.cpp:95]: Remainder of the Orignal source code. I am also not completely sure about its content. 
However the author itself uses the function for more  than just testing :see_no_evil:

File: src/main/include/Subscription.h (hunk 7 above)
Comment (Ti-ger) [reply src/main/include/Subscription.h:70]: This is definitely a follow up. We could use this for the ZAP-Tool integration

File: src/main/CMakeLists.txt (hunk 1 above)
Comment (Ti-ger) [reply src/main/CMakeLists.txt:12]: I introduced this for the "Attribute.h"-header, but it is currently not used. 
I removed it again ;)

**Pull request closed without merging**